This is the diff between the stock CAssetManager and the SafeModeAssetManager I'm currently using. It only changes the publish and getpublishedurl functions, and flattens out directories if the file linked is a single file (i.e. not a directory).

diff -c /tmp/unsafepublish /tmp/safepublish
--- /tmp/unsafepublish	2012-01-10 09:04:01.282259548 -0800
+++ /tmp/safepublish	2012-01-10 09:03:32.670486333 -0800
@@ -9,32 +9,16 @@
 				$dir=$this->hash($hashByName ? basename($src) : dirname($src));
 				$fileName=basename($src);
 				$dstDir=$this->getBasePath().DIRECTORY_SEPARATOR.$dir;
-				$dstFile=$dstDir.DIRECTORY_SEPARATOR.$fileName;
+				$dstFile=$dstDir.'_'.$fileName;
 
 				if($this->linkAssets)
 				{
 					if(!is_file($dstFile))
-					{
-						if(!is_dir($dstDir))
-						{
-							mkdir($dstDir);
-							@chmod($dstDir, $this->newDirMode);
-						}
 						symlink($src,$dstFile);
-					}
 				}
 				else if(@filemtime($dstFile)<@filemtime($src) || $forceCopy)
-				{
-					if(!is_dir($dstDir))
-					{
-						mkdir($dstDir);
-						@chmod($dstDir, $this->newDirMode);
-					}
 					copy($src,$dstFile);
-					@chmod($dstFile, $this->newFileMode);
-				}
-
-				return $this->_published[$path]=$this->getBaseUrl()."/$dir/$fileName";
+				return $this->_published[$path]=$this->getBaseUrl().'/'.$dir.'_'.$fileName;
 			}
 			else if(is_dir($src))
 			{
@@ -47,14 +31,7 @@
 						symlink($src,$dstDir);
 				}
 				else if(!is_dir($dstDir) || $forceCopy)
-				{
-					CFileHelper::copyDirectory($src,$dstDir,array(
-						'exclude'=>$this->excludeFiles,
-						'level'=>$level,
-						'newDirMode'=>$this->newDirMode,
-						'newFileMode'=>$this->newFileMode,
-					));
-				}
+					CFileHelper::copyDirectory($src,$dstDir,array('exclude'=>$this->excludeFiles,'level'=>$level));
 
 				return $this->_published[$path]=$this->getBaseUrl().'/'.$dir;
 			}
@@ -70,11 +47,11 @@
 		if(($path=realpath($path))!==false)
 		{
 			if(is_file($path))
-				return $this->getBaseUrl().'/'.$this->hash($hashByName ? basename($path) : dirname($path)).'/'.basename($path);
+				return $this->getBaseUrl().'/'.$this->hash($hashByName ? basename($path) : dirname($path)).'_'.basename($path);
 			else
 				return $this->getBaseUrl().'/'.$this->hash($hashByName ? basename($path) : $path);
 		}
 		else
 			return false;
 	}
-
+}

Diff finished.  Tue Jan 10 09:04:14 2012
